src/flow/src/batching_mode/state.rs

```diff
@@ -0,0 +1,401 @@
+#[derive(Debug, Clone, Default)]
+pub struct DirtyTimeWindows {
+    /// windows's `start -> end` and non-overlapping
+    /// `end` is exclusive(and optional)
+    windows: BTreeMap<Timestamp, Option<Timestamp>>,
+}
+
+impl DirtyTimeWindows {
+    /// Time window merge distance
+    ///
+    /// TODO(discord9): make those configurable
+    const MERGE_DIST: i32 = 3;
+
+    /// Maximum number of filters allowed in a single query
+    const MAX_FILTER_NUM: usize = 20;
+
+    /// Add lower bounds to the dirty time windows. Upper bounds are ignored.
+    ///
+    /// # Arguments
+    ///
+    /// * `lower_bounds` - An iterator of lower bounds to be added.
+    pub fn add_lower_bounds(&mut self, lower_bounds: impl Iterator<Item = Timestamp>) {
+        for lower_bound in lower_bounds {
+            let entry = self.windows.entry(lower_bound);
+            entry.or_insert(None);
+        }
+    }
+
+    /// Generate all filter expressions consuming all time windows
+    pub fn gen_filter_exprs(
+        &mut self,
+        col_name: &str,
+        expire_lower_bound: Option<Timestamp>,
+        window_size: chrono::Duration,
+        flow_id: FlowId,
+        task_ctx: Option<&BatchingTask>,
+    ) -> Result<Option<datafusion_expr::Expr>, Error> {
+        debug!(
+            "expire_lower_bound: {:?}, window_size: {:?}",
+            expire_lower_bound.map(|t| t.to_iso8601_string()),
+            window_size
+        );
+        self.merge_dirty_time_windows(window_size, expire_lower_bound)?;
+
+        if self.windows.len() > Self::MAX_FILTER_NUM {
+            let first_time_window = self.windows.first_key_value();
+            let last_time_window = self.windows.last_key_value();
+
+            if let Some(task_ctx) = task_ctx {
+                warn!(
+                "Flow id = {:?}, too many time windows: {}, only the first {} are taken for this query, the group by expression might be wrong. Time window expr={:?}, expire_after={:?}, first_time_window={:?}, last_time_window={:?}, the original query: {:?}",
+                task_ctx.flow_id,
+                self.windows.len(),
+                Self::MAX_FILTER_NUM,
+                task_ctx.time_window_expr,
+                task_ctx.expire_after,
+                first_time_window,
+                last_time_window,
+                task_ctx.query
+            );
+            } else {
+                warn!("Flow id = {:?}, too many time windows: {}, only the first {} are taken for this query, the group by expression might be wrong. first_time_window={:?}, last_time_window={:?}",
+                flow_id,
+                self.windows.len(),
+                Self::MAX_FILTER_NUM,
+                first_time_window,
+                last_time_window
+                )
+            }
+        }
+
+        // get the first `MAX_FILTER_NUM` time windows
+        let nth = self
+            .windows
+            .iter()
+            .nth(Self::MAX_FILTER_NUM)
+            .map(|(key, _)| *key);
```

Comment:
IMO we need either (1) print warnings when some time windows are ignored, or (2) handle all dirty windows in multiple runs, with each run has `MAX_FILTER_NUM`

Reply:
> > IMO we need either (1) print warnings when some time windows are ignored, or (2) handle all dirty windows in multiple runs, with each run has `MAX_FILTER_NUM`
> 
> warning is print in above, will refactor to handle all dirty windows in multiple runs too later

also done in multiple runs by setting wait time to zero if still have dirty windows